### ISSUE TITLE
For NMDA, the <intended> datastore returns the expanded template

### ISSUE THREAD

**Issue by kwatsen** [netmod-wg/template-reqs#13]
The <intended> datastore is subject to validation, which means that all the templates need to be expanded/flattened.

Whist templates are expanded in <intended>, it may be that the system consumes the *unexpanded* configuration internally, in order to have a memory-efficient implementation.

**Comment by RobertPeschi** [comment]
Note that templates are also usable in non-NMDA devices: in a non-NMDA device, the device application software would expand the template in its private memory. This expanded data could be made visible to a YANG client by defining a state tree reflecting it.

**Comment by jgcumming** [comment]
I would propose that the expanded version of the template be visible in intended only, this would not include the template itself (which would not be shown).  This view appears to reflect that given during the discussion session at IETF121 from other speakers.

**Comment by RobertPeschi** [comment]
Just to be sure, you mean "_I would propose that the expanded version of the template be visible in intended but not in the running_", right ? 
Yes, this is exactly my point. 
We just have to keep in mind that this statement is specific to NMDA deployments. For non-NMDA deployments (and my take is that many of these will remain so in any foreseeable future), my point is that the expanded version would then be visible in a dedicated R/O state YANG tree (and, again, NOT in the running anyway).  IETF would not standardize a template mechanism only for NMDA deployments, would it ?
  
About your statement "_this would not include the template itself (which would not be shown)_". I don't see why templates would not be part of the intended, too (as an exact copy of templates in the running). My understanding is that anything in the running that remains unchanged should be visible in the intended as well.

**Comment by kwatsen** [comment]
> About your statement "this would not include the template itself (which would not be shown)". I don't see why templates would not be part of the intended, too (as an exact copy of templates in the running). My understanding is that anything in the running that remains unchanged should be visible in the intended as well.

I disagree.  Just like how "inactive" nodes are removed in `<intended>`, so should template definitions be removed in `<intended>`, as they have served their purpose.

See #15

**Comment by RobertPeschi** [comment]
> I disagree. Just like how "inactive" nodes are removed in `<intended>`, so should template definitions be removed in `<intended>`, as they have served their purpose.

Except otherwise decided, the data nodes constituting the template are just "normal" data nodes (e.g. not containing any specific YANG statements - still tbd in case - indicating that their are to be used as template), right ?
Just asking, but then on which basis would a client or a server know which data nodes should be included in the intended and which ones should not ? Any clue ?

**Comment by QiufangMa** [comment]
> I would propose that the expanded version of the template be visible in intended only

intended only? maybe also operational!

**Comment by QiufangMa** [comment]
> This expanded data could be made visible to a YANG client by defining a state tree reflecting it.

I thought another alternative is to define an RPC for the expanded data. But I am unsure template solution should include non-NMDA or not.

**Comment by RobertPeschi** [comment]
> (...). But I am unsure template solution should include non-NMDA or not.

A template solution valid only for NMDA would be too limitative in my view

**Comment by derajara** [comment]
> > About your statement "this would not include the template itself (which would not be shown)". I don't see why templates would not be part of the intended, too (as an exact copy of templates in the running). My understanding is that anything in the running that remains unchanged should be visible in the intended as well.
> 
> I disagree. Just like how "inactive" nodes are removed in `<intended>`, so should template definitions be removed in `<intended>`, as they have served their purpose.
> 
> See #15

Yes, that's right, inactive nodes can be removed in intended(but then we need to mark such nodes/configuration as inactive, not sure if today we have a standard way to mark certain nodes/configuration as inactive), but here, the templates are not inactive, but active configuration (meaning, there is an active edit-config from a client to create a template irrespective of whether they are referred by any consumers or not), so I don't see a reason why this should not be in the intended.

**Comment by jsterne** [comment]
IMO a (non-expanded) template is a "pre transformation" piece of config (like inactive config) that should be in < running > but not in < intended >.

For < operational > I've created a separate issue #17

**Comment by RobertPeschi** [comment]
>IMO a (non-expanded) template is a "pre transformation" piece of config (like inactive config) that should be in < running > but not in < intended >.
'Trying to understand here - what would hurt if a non expanded template was in the intended DS ? 

The reason why I think that excluding by principle non-expanded templates from the intended DS is too restrictive:
- The intended DS is the data that the server presents to the device application SW, right ?
- In draft-rajaram-yang-config-template-framework, it is the applications SW in the device that will expand the template. 
- Thus to do so the not-yet-expanded template must naturally be available in the tentative DS.

Since we don't know yet the details of the ultimate template mechanism, I think it is wise not to arbitrarily restrict our possibilities. Hence if it does not hurt, I'd keep unexpanded templates in the tentative DS.
